archive/android_sparse.ksy

```diff
@@ -0,0 +1,93 @@
+meta:
+  id: android_sparse
+  title: Android Sparse image
+  file-extension: img
+  tags:
+    - archive
+    - android
+  license: CC0-1.0
+  endian: le
+doc: |
+    The Android sparse format is a format to more efficiently store files
+    for for example firmware updates to save on bandwidth. Files in sparse
+    format first have to be converted back to their original format.
+
+    A tool to create images for testing can be found in the Android source code tree:
+
+    <https://android.googlesource.com/platform/system/core/+/7b444f0/libsparse> - `img2simg.c`
+
+    Note: this is not the same as the Android sparse data image format.
+doc-ref: https://android.googlesource.com/platform/system/core/+/7b444f0/libsparse/sparse_format.h
+seq:
+  - id: img_header
+    type: header
+  - id: img_header_entries
+    type: image_header_entry
+    repeat: expr
+    repeat-expr: img_header.total_chunks
+types:
+  header:
+     seq:
+       - id: magic
+         contents: [0x3a, 0xff, 0x26, 0xed]
+       - id: version
+         type: version
+       - id: file_header_size
+         -orig-id: file_hdr_sz
+         type: u2
+         doc: size of file header, should be 28
+       - id: chunk_header_size
+         -orig-id: chunk_hdr_sz
+         type: u2
+         doc: size of chunk header, should be 12
+       - id: block_size
+         -orig-id: blk_sz
+         type: u4
+         doc: block size in bytes, multiple of 4
+       - id: total_blocks
+         -orig-id: total_blks
+         type: u4
+         doc: blocks in the original data
+       - id: total_chunks
+         -orig-id: total_chunks
+         type: u4
+       - id: checksum
+         -orig-id: image_checksum
+         type: u4
+         doc: CRC32 checksum of the original data
+  image_header_entry:
+    seq:
+      - id: header
+        type: fixed_header
+      - id: body
+        size: header.total_size - header._sizeof
```

Comment:
I've just stumbled upon a file called [`oem_ru_base.img`](https://herelinkfw.cubepilot.org/tools/oem_ru_base.img) from https://docs.cubepilot.org/user-guides/herelink/herelink-user-guides/oem-image-setup:

> ### [Steps](https://docs.cubepilot.org/user-guides/herelink/herelink-user-guides/oem-image-setup#steps)
> 
> Download and unzip following files:
> 
> * (...)
> * Remote Unit: [oem_ru_base.img](https://herelinkfw.cubepilot.org/tools/oem_ru_base.img)

Parsing this sparse file based on the `.ksy` spec fails on the very first `imageHeader` entry with an error:

```
RangeError: Invalid typed array length: -11
```

After doing a temporary modification:

```diff
--- 1/android_sparse@old.ksy
+++ 2/android_sparse@new.ksy
@@ -24,7 +24,8 @@ seq:
   - id: img_header_entries
     type: image_header_entry
     repeat: expr
-    repeat-expr: img_header.total_chunks
+    repeat-expr: 1
+    # repeat-expr: img_header.total_chunks
 types:
   header:
      seq:
@@ -59,8 +60,8 @@ types:
     seq:
       - id: header
         type: fixed_header
-      - id: body
-        size: header.total_size - header._sizeof
+      # - id: body
+      #   size: header.total_size - header._sizeof
     types:
       fixed_header:
         seq:
```

the following structure appears:

```js
{
  "imgHeader": {
    "magic": [58, 255, 38, 237],
    "version": {
      "major": 1,
      "minor": 0
    },
    "fileHeaderSize": 32,
    "chunkHeaderSize": 16,
    "blockSize": 4096,
    "totalBlocks": 76800,
    "totalChunks": 26,
    "checksum": 0
  },
  "imgHeaderEntries": [
    {
      "header": {
        "chunkType": 0,
        "reserved": 0,
        "chunkSize": 0xcac1,
        "totalSize": 1
      }
    }
  ]
}
```

So the parsing errorred on `totalSize` being 1, which makes `header.total_size - header._sizeof` negative (`header.total_size - header._sizeof = 1 - 12 = -11`). And `chunkType` and `chunkSize` have clearly swapped places 🤷

The `simg2img` has no problem with this file though, so it must be readable somehow. I will investigate further.

Anyway, I think this is a great example why we need sample files, and not only from one source or generator, but ideally diverse from many different sources. Implementing a format in a `.ksy` file according to some existing parser or specification is one thing, but whether it actually is capable of parsing real-life files is a totally different thing, and there is no other way to find out than to collect a variety of samples and test them. Not always are specifications compatible with how the format is actually used in the wild. Moreover, omissions and errors are common among both format descriptions and parsers/generators.

Reply:
I used simg2img and then img2simg and getting a significantly smaller result, which then is parsed correctly. I looked at the two, and also ran simg_dump.py ( https://github.com/anestisb/android-simg2img/blob/master/simg_dump.py ) which complains about the file:

```
$ python2 simg_dump.py oem_ru_base.img
simg_dump.py: oem_ru_base.img: The file header size was expected to be 28, but is 32.
```

and indeed, the header size (`file_header_size`) is 32, whereas in "normal" sparse image files it is supposed to be 28.

Reply:
Added four padding bytes to fill the header to 32 and the file is parsed correctly. So the problem is `file_header_size`. I can see a few solutions:

1. adding padding bytes depending on the size of the header
2. splitting the header in two parts
3. making `image_header_entries` an instance starting at pos `file_header_size` (probably after some sanity checks)

My personal preference would be 1.